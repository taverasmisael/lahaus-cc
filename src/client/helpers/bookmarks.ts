import { Bookmark } from '@/shared/bookmark.interface'
import { RealEstate } from '@/shared/real-state.interface'
import { ResponseBookmarks } from '@/shared/response-bookmarks.interface'

export const normalizeResponse = (res: ResponseBookmarks): Bookmark[] => {
  const { data, included } = res

  const bookmarks = data.map((bm) => {
    const realEstates = bm.attributes.real_estate_ids.reduce((prev, id) => {
      const realEstate = included.find((state) => state.id === id.toString())
      return realEstate ? [...prev, realEstate] : prev
    }, [] as RealEstate[])

    return {
      id: bm.id,
      name: bm.attributes.name,
      realEstates,
    }
  })

  return bookmarks
}
