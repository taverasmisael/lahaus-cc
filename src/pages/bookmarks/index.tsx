// TODO: IMPLEMENT. Just redirect to the home page for now
function BookmarksPage() {
  return <div />
}

export const getServerSideProps = () => {
  return {
    redirect: {
      destination: '/',
      permanent: false,
    },
  }
}

export default BookmarksPage
