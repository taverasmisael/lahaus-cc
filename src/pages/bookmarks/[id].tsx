// next/image was crashing the app on resize for some reason ¯\_(ツ)_/¯
/* eslint-disable @next/next/no-img-element */

import type { NextPage, GetServerSideProps } from 'next'
import Link from 'next/link'
import AppShell from '@/layouts/AppShell'
import { ResponseBookmarks } from '@/shared/response-bookmarks.interface'
import { normalizeResponse } from '@/client/helpers/bookmarks'
import { RealEstate } from '@/shared/real-state.interface'

const BookmarkPage: NextPage<BookmarkPageProps> = ({ realEstates, name }) => {
  return (
    <AppShell title={`Guardados en "${name}"`}>
      <div className="container px-4 mx-auto">
        <header className="mb-14">
          <h1 className="text-center mb-4">{name}</h1>
          <Link href="/bookmarks">
            <a className="text-center text-sm text-gray-400 block hover:text-copy">Volver a los marcadores</a>
          </Link>
        </header>
        {!realEstates.length && (
          <h2 className="text-center">Agrega algun favorito</h2>
        )}

        <div className="grid gap-4 md:grid-cols-3 2xl:grid-cols-4">
          {realEstates.map((re) => (
            <div key={re.id} className="rounded-md grid overflow-hidden shadow-sm hover:shadow-md transition-shadow duration-300">
              <img
                src={re.attributes.gallery_urls[0]}
                alt={re.attributes.name}
                className="w-full object-cover h-60 "
              />
              <div className="p-4 grid">
                <h3 className="text-xl mb-4">{re.attributes.name}</h3>
                <dl className="flex gap-2 text-sm">
                  <dt className="font-bold">Habitaciones</dt>
                  <dd>{re.attributes.bedroom[0]}</dd>
                  <dt className="font-bold ml-2">Baños</dt>
                  <dd>{re.attributes.bathroom[0]}</dd>
                </dl>

                <a
                  href="#"
                  className="inline-block w-full mt-8 p-4 text-center bg-secondary text-gray-100 hover:text-white rounded-sm font-semibold text-lg hover:shadow-md"
                >
                  Ver detalles
                </a>
              </div>
            </div>
          ))}
        </div>
      </div>
    </AppShell>
  )
}

export const getServerSideProps: GetServerSideProps = async (ctx) => {
  try {
    const url = process.env.NEXT_PUBLIC_API_URL

    const id = ctx.params?.id
    if (!id) {
      throw new Error('id is required')
    }
    if (!url) {
      throw new Error('Missing API URL')
    }

    const res = await fetch(url)

    if (!res.ok) {
      throw new Error(`Could not fetch ${url}`)
    }

    const data: ResponseBookmarks = await res.json()
    const result = normalizeResponse(data).find((item) => item.id === id)

    if (!result) {
      throw new Error(`Could not find ${id}`)
    }

    return {
      props: {
        name: result.name,
        realEstates: result.realEstates,
      },
    }
  } catch (e) {
    return {
      redirect: {
        destination: '/',
        permanent: false,
      },
    }
  }
}

interface BookmarkPageProps {
  name: string
  realEstates: RealEstate[]
}

export default BookmarkPage
