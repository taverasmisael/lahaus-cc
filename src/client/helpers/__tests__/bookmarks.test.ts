import { RealEstateAttributes } from '@/shared/real-state.interface'
import { ResponseBookmarks } from '@/shared/response-bookmarks.interface'
import { normalizeResponse } from '../bookmarks'

describe('helpers/bookmarks', () => {
  describe('normalizeResponse', () => {
    test('can normalize an API response', () => {
      const response: ResponseBookmarks = {
        data: [{ attributes: { name: 'TEST', real_estate_ids: [3] }, id: '1' }],
        included: [
          { id: '3', attributes: { name: 'TEST' } as RealEstateAttributes },
        ],
      }

      expect(normalizeResponse(response)).toMatchSnapshot()
    })

    test('can normalize an API response even with no real estate matches', () => {
      const response: ResponseBookmarks = {
        data: [{ attributes: { name: 'TEST', real_estate_ids: [] }, id: '1' }],
        included: [
          { id: '3', attributes: { name: 'TEST' } as RealEstateAttributes },
        ],
      }
      expect(normalizeResponse(response)).toMatchSnapshot()
    })
    test('ignore not found real estates ids', () => {
      const response: ResponseBookmarks = {
        data: [{ attributes: { name: 'TEST', real_estate_ids: [2] }, id: '1' }],
        included: [
          { id: '3', attributes: { name: 'TEST' } as RealEstateAttributes },
        ],
      }
      expect(normalizeResponse(response)).toMatchSnapshot()
    })
  })
})
