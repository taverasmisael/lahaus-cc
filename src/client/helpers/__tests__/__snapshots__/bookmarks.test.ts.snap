// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers/bookmarks normalizeResponse can normalize an API response 1`] = `
Array [
  Object {
    "id": "1",
    "name": "TEST",
    "realEstates": Array [
      Object {
        "attributes": Object {
          "name": "TEST",
        },
        "id": "3",
      },
    ],
  },
]
`;

exports[`helpers/bookmarks normalizeResponse can normalize an API response even with no real estate matches 1`] = `
Array [
  Object {
    "id": "1",
    "name": "TEST",
    "realEstates": Array [],
  },
]
`;

exports[`helpers/bookmarks normalizeResponse ignore not found real estates ids 1`] = `
Array [
  Object {
    "id": "1",
    "name": "TEST",
    "realEstates": Array [],
  },
]
`;
